### PR TITLE
ConcurrentPercolator: remove the pre rollback 

### PR DESCRIPTION
**Pull request by cwen0**
This pr is to verify this https://github.com/pingcap/tikv/pull/3290 

When writing `rollback` to `key_write`, the previous `rollback` with the same key will be deleted first, and the ts of the deleted `rollback` must be less than or equal to the `rollback` to be written.  

PTAL @foreverbell @siddontang @GregoryIan @zhangjinpeng1987

### CODE REVIEW

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 1]
```diff
@@ -74,7 +73,7 @@ hasLockEQ(k, ts) ==
   \E l \in key_lock[k] : l.ts = ts
 
 \* Returns TRUE if a lock can be cleanup up.
-\* A lock can be cleaned up iff its ts is less than or equal to $ts$.
+\* A lock can be cleaned up if its ts is less than or equal to $ts$.
```
Comment (siddontang) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: iff means if and only if 

it is fine here

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 2]
```diff
@@ -102,8 +101,41 @@ checkSnapshotIsolation(k, commit_ts) ==
...
     key_si' = [key_si EXCEPT ![k] = FALSE]
   ELSE
-    UNCHANGED <<key_si>>
-
+    UNCHANGED <<key_si>> 
+
+\* Returns index of $e$ in $seq$ 
+getIndex(seq, e) == CHOOSE n \in DOMAIN seq : seq[n] = e  
+
+\* Returns the write whose $ts$ is largest
```
Comment (siddontang) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: em, the last one in the write sequence must have the latest commit ts

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 3]
```diff
@@ -102,8 +101,41 @@ checkSnapshotIsolation(k, commit_ts) ==
...
+getWriteWithMaxTS(w) == CHOOSE x \in w : \A y \in w : x.ts >= y.ts
+
+\* Returns the writes with ts less or equal than $ts$ 
+findWriteLessEqualTS(kw, ts) ==
+  {w \in Range(kw) : (w.ts <= ts)}  
+
+\* Deletes element whose index in $seq$ equal to $index$
+deleteElement(seq, index)==
+  [i \in 1..(Len(seq)-1)|->IF i<index THEN seq[i]ELSE seq[(i+1)]]
```
Comment (siddontang) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: `THEN seq[i] ELSE`

add a blank

**Comment by IANTHEREAL** [comment]
I think we shouldn't delete rollbacl record.  If some one rollbacks a txn, it's what happened,  we shouldn't delete it in the verification, otherwise we lose a real record and cannot verify it accurately.    Just mark it deleted.

File: ConcurrentPercolator/ConcurrentPercolator.tla (hunk 2 above)
Comment (cwen0) [reply ConcurrentPercolator/ConcurrentPercolator.tla]: We did not verify that the write sequence must be ordered

Comment (cwen0) [reply ConcurrentPercolator/ConcurrentPercolator.tla]: we just guarantee that the `commit` in `key_write` is ordered, but the `rollback` is not guaranteed

Comment (siddontang) [reply ConcurrentPercolator/ConcurrentPercolator.tla]: why?

Comment (IANTHEREAL) [reply ConcurrentPercolator/ConcurrentPercolator.tla]: we delete rollback record earlier, now we don't have this problem

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 4]
```diff
@@ -61,11 +61,10 @@ vars == <<next_ts, client_vars, key_vars>>
 
 --------------------------------------------------------------------------------
 Range(m) ==
-  {m[i] : i \in DOMAIN m}
+  {m[i] : i \in DOMAIN m}
```
Comment (IANTHEREAL) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: why are there spaces at the end?

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 5]
```diff
@@ -84,16 +83,30 @@ hasStaleLock(k, ts) ==
...
 findWriteWithStartTS(k, ts) ==
-  {w \in Range(key_write[k]) : (w.type = "write" /\ w.start_ts = ts)}
-
+  {w \in Range(key_write[k]) : (w.type = "write" /\ w.start_ts = ts)}      
+   
 \* Returns the writes with commit_ts equals to $ts$.
 findWriteWithCommitTS(k, ts) ==
   {w \in Range(key_write[k]) : (w.type = "write" /\ w.ts = ts)}
 
 \* Returns TRUE if there is a rollback for key $k$ at timestamp $ts$.
```
Comment (IANTHEREAL) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: \* Returns TRUE if there is a rollback for key $k$ that is not deleted logically at timestamp $ts $.

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 6]
```diff
@@ -102,8 +115,51 @@ checkSnapshotIsolation(k, commit_ts) ==
...
+
+\* Returns index of $e$ in $seq$ 
+getIndex(seq, e) == CHOOSE n \in DOMAIN seq : seq[n] = e  
+
+\* Returns the write whose $ts$ is largest
+getWriteWithMaxTS(w) == CHOOSE x \in w : \A y \in w : x.ts >= y.ts
+
+\* Returns the writes with ts less or equal than $ts$ 
+findWriteLessEqualTS(kw, ts) ==
+  {w \in Range(kw) : (w.ts <= ts /\ IF (w.type = "rollback" /\ w.is_deleted = TRUE) THEN FALSE ELSE TRUE)}
```
Comment (IANTHEREAL) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: refine it, key werite  is ordered.

Comment (IANTHEREAL) [reply ConcurrentPercolator/ConcurrentPercolator.tla]: anyway why use `if then else`, just use  `(w.type = "write" \/ w.is_deleted = FALSE)`

**Comment by cwen0** [comment]
PTAL again @GregoryIan @siddontang, Thanks!

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 7]
```diff
@@ -90,10 +89,24 @@ findWriteWithStartTS(k, ts) ==
 findWriteWithCommitTS(k, ts) ==
   {w \in Range(key_write[k]) : (w.type = "write" /\ w.ts = ts)}
 
-\* Returns TRUE if there is a rollback for key $k$ at timestamp $ts$.
+\* Returns TRUE if there is a rollback for key $k$ that is not deleted logically at timestamp $ts$.
 hasRollback(k, ts) ==
+  {r \in Range(key_write[k]) : (r.type = "rollback" /\ r.is_deleted = FALSE /\ r.ts = ts )} # {}
+
+\* Returns TRUE if there is a rollback for key $k$ at timestamp $ts$
+hasRollbackIgnoreDeleted(k, ts) ==
```
Comment (IANTHEREAL) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: it should be call `hasRollback

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 8]
```diff
@@ -104,6 +117,38 @@ checkSnapshotIsolation(k, commit_ts) ==
   ELSE
     UNCHANGED <<key_si>>
 
+\* Returns the index of the first write with ts less or equal than $ts$
+getWritesLessEqualTS(ws, ts) == SelectSeq(ws, LAMBDA w: (w.ts <= ts /\ w.is_deleted = FALSE))
+
+\* Logical deletes element whose index in $seq$ equal to $index$
```
Comment (IANTHEREAL) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: deletes element whose index in $seq$ equal to $index$ logically

**Review by IANTHEREAL — COMMENTED** [review]
Rest LGTM

**Comment by cwen0** [comment]
@GregoryIan PTAL

**Comment by IANTHEREAL** [comment]
LGTM

**Comment by cwen0** [comment]
@siddontang PTAL

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 9]
```diff
@@ -104,6 +117,45 @@ checkSnapshotIsolation(k, commit_ts) ==
...
+      ELSE
+          w
+
+\* Writes rollback record to key write.
+\* If the rollback with the ts equal to $ts$ exists and is marked as deleted,
+\* which will be overwritten by the new rollback.
+writeRollback(ws, ts) ==
+  IF hasRollbackWithDeleted(ws, ts)
+  THEN
+     [i \in 1..Len(ws)|->IF isRollbackWithDeleted(ws[i], ts) THEN ws[i] ELSE [ts |-> ws[i].ts, type |-> ws[i].type, is_deleted |-> TRUE]]
```
Comment (siddontang) [inline ConcurrentPercolator/ConcurrentPercolator.tla]: here you mean if there is a rollback with Deleted, we will set other rollbacks to Deleted too?

Comment (cwen0) [reply ConcurrentPercolator/ConcurrentPercolator.tla]: we have guaranteed that there is only one in the key_write with the same `ts`. we just overwrite this `rollback`.  O, no, my fault, we should set `is_deleted |-> FALSE`

**Comment by cwen0** [comment]
@siddontang PTAL again

**Comment by cwen0** [comment]
@siddontang PTAL again

**Review by siddontang — APPROVED** [review]
LGTM

**Pull request merged**